rbd: fix bug where volumereplication promote/demote would fail

SkalaNetworks opened the pull request:
If the source and destination clusters use different clusterIDs, promoting/demoting a volumereplication would fail due to the clusterIDmapping logic not running.

This fix enables clusterIDmapping during volume generation to fix it.

<!-- Please take a look at our [Contributing](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#Code-contribution-workflow)
documentation before submitting a Pull Request!
Thank you for contributing to ceph-csi! -->

# Describe what this PR does #

This PR fixes https://github.com/ceph/ceph-csi/issues/5529. 

Importing a PVC to run a disaster recovery procedure would fail if the clusterID of the 2 clusters were different.

For example, if Ceph-CSI is installed through Rook, the clusterID is that of the deployment namespace. There's already a mapping logic that takes care of that, but in the context of promotion/demotions from primary/secondary, it simply wasn't called.

## Is there anything that requires special attention ##

Is the change backward compatible? Yes

Are there concerns around backward compatibility? No

## Related issues ##

Fixes: https://github.com/ceph/ceph-csi/issues/5529


**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow
  guidelines in the [developer
  guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull
  Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release
  notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md)
  updated with breaking and/or notable changes for the next major release.
* [ ] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [ ] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

SkalaNetworks committed:
rbd: fix bug where volumereplication promote/demote would fail

If the source and destination clusters use different clusterIDs,
promoting/demoting a volumereplication would fail due to the
clusterIDmapping logic not running.

This fix enables clusterIDmapping during volume
generation to fix it.

Signed-off-by: Skala Networks <contact@skala.network>

Rakshith-R commented:
@SkalaNetworks 
I think the correct approach would be to 
- Create a error type for the following in https://github.com/ceph/ceph-csi/blob/ef48973007a3a650224f9797a418c0799216b688/internal/util/errors.go
https://github.com/ceph/ceph-csi/blob/ef48973007a3a650224f9797a418c0799216b688/internal/util/csiconfig.go#L86

- Use that error in the following function to determine if volume generation should be retried
https://github.com/ceph/ceph-csi/blob/dc1f3d7c8fc46e66387cf3281654211719e34e9c/internal/rbd/rbd_util.go#L1264-L1273

Can you please try this? 

I think we may have missed this issue since we use `rook-ceph` as the clusterID most of the time.

SkalaNetworks commented:
@Rakshith-R here you go, I also fixed the bug in VolumeGroupReplications and adjusted the tests.

Note: there's a large amount of code duplication between the clusterIDmapping in function `GetMonsAndClusterID` and in `GenVolFromVolID`. 

`GetMonsAndClusterID` is capable of resolving mapping using a single boolean. `GenVolFromVolID` runs pretty much the same logic but without using the same code to do the mapping.


NOTE: Tested on a cluster, I could promote/demote just fine

SkalaNetworks commented:
Working on running make go-lint but I'm getting

```
../internal/rbd/errors/errors.go:24:9: could not import github.com/ceph/go-ceph/rbd (-: # github.com/ceph/go-ceph/rbd
vendor/github.com/ceph/go-ceph/rbd/encryption_load2.go:97:9: could not determine what C.rbd_encryption_load2 refers to
vendor/github.com/ceph/go-ceph/rbd/encryption_load2.go:20:28: could not determine what C.rbd_encryption_spec_t refers to) (typecheck)
	librbd "github.com/ceph/go-ceph/rbd"
```

Nikhil-Ladha commented:
> Working on running make go-lint but I'm getting
> 
> ```
> ../internal/rbd/errors/errors.go:24:9: could not import github.com/ceph/go-ceph/rbd (-: # github.com/ceph/go-ceph/rbd
> vendor/github.com/ceph/go-ceph/rbd/encryption_load2.go:97:9: could not determine what C.rbd_encryption_load2 refers to
> vendor/github.com/ceph/go-ceph/rbd/encryption_load2.go:20:28: could not determine what C.rbd_encryption_spec_t refers to) (typecheck)
> 	librbd "github.com/ceph/go-ceph/rbd"
> ```

Are you trying to build with a different version of ceph image?

SkalaNetworks commented:
@Nikhil-Ladha I ran make containerized-build && make go-lint

I also installed librbd/rados libraries because the go-lint was trying to build the code and I didn't have a few headers

Description:	Ubuntu 22.04.5 LTS
apt install librados-dev
apt install librbd-dev

Nikhil-Ladha commented:
> @Nikhil-Ladha I ran make containerized-build && make go-lint
> 
> I also installed librbd/rados libraries because the go-lint was trying to build the code and I didn't have a few headers
> 
> Description: Ubuntu 22.04.5 LTS apt install librados-dev apt install librbd-dev

Ah, possibly the version of headers you have in the local machine is different that the one we have in the go-ceph version linked to ceph-csi.
Please try this command: `make containerized-test TARGET=go-lint` instead.

SkalaNetworks commented:
Thanks, it seems to have worked just fine

iPraveenParihar reviewed (approved):
Thanks!

Nikhil-Ladha reviewed (approved):
/lgtm
Thanks.

Rakshith-R reviewed (approved):
Thanks !

SkalaNetworks commented:
Can we get this backported into the latest version?

iPraveenParihar commented:
@Mergifyio queue
